### DISCUSSION TITLE
Support for azure workload identity

### DISCUSSION THREAD

**Discussion by tianqi-z** [sqlalchemy/sqlalchemy#10530] [Usage Questions] [answered]
### Describe the use case

Hi team, 
I have an aks with workload identity enabled. https://learn.microsoft.com/en-us/azure/aks/workload-identity-overview?tabs=python
I have a user assigned managed identity created for WI and this user is the aad admin for the sql server. I have a pod on the AKS trying to connect to the same db using different type of connection strings. (ADO, JDBC, SQLALCHEMY PYODBC). The ADO and JDBC connections from the same pod were successful using workload identity, which proved that I had the workload identity setup correctly. Is workload identity supported or could you help on identifying the issues? Thanks.

### Databases / Backends / Drivers targeted

Azure SQL
ODBC Driver 18 for SQL Server

### Example Use

Part of my docker file for the pod:
```
FROM registry.access.redhat.com/ubi8/ubi-minimal:8.8 as ubi
WORKDIR /bin
ENV ACCEPT_EULA=Y
RUN curl https://packages.microsoft.com/config/rhel/8/prod.repo > /etc/yum.repos.d/mssql-release.repo
RUN microdnf -y install unixODBC openssl libtool-ltdl dotnet-runtime-6.0 java-1.8.0-openjdk-headless python3 python3-pip unixODBC-devel gcc-c++ python3-devel krb5-workstation krb5-libs mssql-tools
RUN microdnf -y download msodbcsql18
RUN microdnf -y install msodbcsql18-*.x86_64
RUN microdnf clean all
RUN pip3 install sqlalchemy pyodbc setuptools-rust azure.identity 
```
Part of my code
```
def connect_sql(connection_string):
        engine = sqlalchemy.create_engine(connection_string)
        sqlalchemy.event.listen(target=engine, identifier="do_connect", fn=provide_token)
        connection = engine.connect()

def provide_token(dialect, conn_rec, cargs, cparams):
    cargs[0] = cargs[0].replace(";Trusted_Connection=Yes", "")
    azure_federated_token_file = os.getenv('AZURE_FEDERATED_TOKEN_FILE', '')
    f = open(azure_federated_token_file, 'rb')
    raw_token = f.read()
    token_struct = struct.pack(f"<I{len(raw_token)}s", len(raw_token), raw_token)
    # apply it to keyword arguments
    cparams["attrs_before"] = {SQL_COPT_SS_ACCESS_TOKEN: token_struct}
```
connection string: 
mssql+pyodbc://@servername.database.windows.net:1433/dbname?driver=ODBC+Driver+18+for+SQL+Server

Error I got:
[Microsoft][ODBC Driver 18 for SQL Server][SQL Server]Login failed for user 'token-identified principal'. (18456) (SQLDriverConnect)\")\n(Background on this error at: https://sqlalche.me/e/14/rvf5)

### Additional context

@zzzeek @CaselIT could you help take a look on this issue? Thanks.

**Comment by CaselIT** [comment]
Hi,

Can you confirm that connecting using pyodbc directly works as expected?

Cc @gordthompson

**Reply by CaselIT** [reply]
Looking at the example in the docs https://docs.sqlalchemy.org/en/20/dialects/mssql.html#connecting-to-databases-with-access-tokens

You are missing an encode as utf16le. Not sure if it's the issue but it may be?

**Reply by tianqi-z** [reply]
Hi @CaselIT , Thanks for looking! 
I updated it with below but still got the same error. 
```
f = open(azure_federated_token_file, 'rb')
token_contents = f.read()
decoded_contents = token_contents.decode('utf-8')
raw_token =decoded_contents.encode('utf-16-le')
```

**Reply by CaselIT** [reply]
Using pyodbc directly has better success?

**Reply by tianqi-z** [reply]
@CaselIT Unfortunately, we can only use sqlalchemy instead of pyodbc. Two possible issues:
1. is my connection string correct: mssql+pyodbc://@servername.database.windows.net:1433/dbname?driver=ODBC+Driver+18+for+SQL+Server
2.  what is the proper way to process the token (the one workload identity created, stored in env var AZURE_FEDERATED_TOKEN_FILE)?
Thanks!

**Reply by CaselIT** [reply]
> Unfortunately, we can only use sqlalchemy instead of pyodbc

note that sqlalchemy uses pyodbc, so from the db point of view there is no difference.

By try using pyodbc directly I mainly meant use it to establish the connection, then pass provide it to sqlalchemy: https://docs.sqlalchemy.org/en/20/core/engines.html#fully-replacing-the-dbapi-connect-function

this way you can use a fully custom pyodbc connect call and pass it to sqlalchemy

**Reply by tianqi-z** [reply]
@CaselIT pyodbc failed with same error. It might be related to how I handle the token. Any suggestions?

**Reply by gordthompson** [reply]
If you have repro code using plain pyodbc that fails then you might consider asking about it at

https://github.com/mkleehammer/pyodbc/discussions

**Reply by tianqi-z** [reply]
I made it work by creating the token similar with https://github.com/Azure/azure-workload-identity/blob/main/examples/msal-python/main.py#L15.

**Comment by tianqi-z** [comment, marked as answer]
Generate the token by: https://github.com/Azure/azure-workload-identity/blob/main/examples/msal-python/main.py#L15
